### PR TITLE
nettests: randomize input

### LINKED ISSUE #1023: Add support for shuffling the order of input before running them
Since we want to enforce a time based limit on MK tests for mobile app, we want to be shuffling the inputs before we run them.

As a first implementation we should do this by shuffling it in memory.

### PR DESCRIPTION
**Pull request by bassosimone**
Closes #1023.

### CODE REVIEW

File: src/libmeasurement_kit/nettests/runnable.cpp [hunk 1]
```diff
@@ -263,30 +260,31 @@ void Runnable::begin(Callback<Error> cb) {
...
                     }
-                    // See http://stackoverflow.com/questions/5750485
-                    //  and http://stackoverflow.com/questions/28331017
-                    input_generator->clear();
-                    input_generator->seekg(0);
-
+                    // See http://en.cppreference.com/w/cpp/algorithm/shuffle
+                    std::random_device rd;
+                    std::mt19937 g(rd());
+                    std::shuffle(inputs.begin(), inputs.end(), g);
```
Comment (hellais) [inline src/libmeasurement_kit/nettests/runnable.cpp]: Can we make the shuffling of inputs something that is enabled optionally based on a flag? (Maybe the flag could be on by default, but it should be possible to change the behaviour).

Comment (bassosimone) [inline src/libmeasurement_kit/nettests/runnable.cpp]: Yeah, sure, let me cook a diff for that!

Comment (bassosimone) [inline src/libmeasurement_kit/nettests/runnable.cpp]: @hellais, I've implemented it in 893043e, thanks!

**Comment by hellais** [comment]
The diff looks good to me, just had the comment on #1043.

I will do some testing locally and then we can merge this.

**Comment by hellais** [comment]
I guess there should be some unittests for this feature though.

**Comment by bassosimone** [comment]
Added regress test. Example run:

```
./test/nettests/runnable -s "Make sure that 'randomize_input' works"
[!] geoip_country_path is not set
[!] geoip_asn_path is not set

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lt-runnable is a Catch v1.5.9 host application.
Run with -? for options

-------------------------------------------------------------------------------
Make sure that 'randomize_input' works
  In the common case
-------------------------------------------------------------------------------
test/nettests/runnable.cpp:167
...............................................................................

test/nettests/runnable.cpp:199: 
PASSED:
  REQUIRE( result != expect )
with expansion:
  { "cyber.law.harvard.edu", "duckduckgo.com", "www.emule.com",
  "creativecommons.org", "ooni.nu", "neubot.org", "nmap.org", "netflix.com",
  "torproject.org", "archive.org" }
  !=
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }

[!] geoip_country_path is not set
[!] geoip_asn_path is not set
-------------------------------------------------------------------------------
Make sure that 'randomize_input' works
  When the user does not want input to be shuffled
-------------------------------------------------------------------------------
test/nettests/runnable.cpp:167
...............................................................................

test/nettests/runnable.cpp:221: 
PASSED:
  REQUIRE( result == expect )
with expansion:
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }
  ==
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }

===============================================================================
All tests passed (2 assertions in 1 test case)


```

**Comment by bassosimone** [comment]
I started wondering whether the test could fail because a random permutation of a vector is not totally unlikely to be equal to the original vector, as such I'll commit soon a more robust test.

**Comment by bassosimone** [comment]
Example test run output:

```
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lt-runnable is a Catch v1.5.9 host application.
Run with -? for options

-------------------------------------------------------------------------------
Make sure that 'randomize_input' works
  In the common case
-------------------------------------------------------------------------------
test/nettests/runnable.cpp:167
...............................................................................

test/nettests/runnable.cpp:210: 
PASSED:
  REQUIRE( run(true) != run(true) )
with expansion:
  { "nmap.org", "netflix.com", "www.emule.com", "torproject.org",
  "creativecommons.org", "neubot.org", "duckduckgo.com", "archive.org", "ooni
  .nu", "cyber.law.harvard.edu" }
  !=
  { "duckduckgo.com", "ooni.nu", "creativecommons.org", "cyber.law.harvard.edu"
  , "torproject.org", "neubot.org", "archive.org", "netflix.com", "nmap.org",
  "www.emule.com" }

[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
test/nettests/runnable.cpp:211: 
PASSED:
  REQUIRE( run(true) != run(true) )
with expansion:
  { "duckduckgo.com", "netflix.com", "cyber.law.harvard.edu", "torproject.org",
  "archive.org", "creativecommons.org", "neubot.org", "www.emule.com", "ooni
  .nu", "nmap.org" }
  !=
  { "cyber.law.harvard.edu", "www.emule.com", "duckduckgo.com", "torproject
  .org", "creativecommons.org", "neubot.org", "nmap.org", "archive.org",
  "netflix.com", "ooni.nu" }

[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
test/nettests/runnable.cpp:212: 
PASSED:
  REQUIRE( run(true) != run(true) )
with expansion:
  { "archive.org", "ooni.nu", "cyber.law.harvard.edu", "nmap.org", "neubot.org"
  , "www.emule.com", "duckduckgo.com", "netflix.com", "creativecommons.org",
  "torproject.org" }
  !=
  { "creativecommons.org", "ooni.nu", "torproject.org", "cyber.law.harvard.edu"
  , "neubot.org", "nmap.org", "archive.org", "www.emule.com", "netflix.com",
  "duckduckgo.com" }

[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
-------------------------------------------------------------------------------
Make sure that 'randomize_input' works
  When the user does not want input to be shuffled
-------------------------------------------------------------------------------
test/nettests/runnable.cpp:167
...............................................................................

test/nettests/runnable.cpp:218: 
PASSED:
  REQUIRE( vec == run(false) )
with expansion:
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }
  ==
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }

test/nettests/runnable.cpp:219: 
PASSED:
  REQUIRE( vec == expect )
with expansion:
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }
  ==
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }

[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
test/nettests/runnable.cpp:218: 
PASSED:
  REQUIRE( vec == run(false) )
with expansion:
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }
  ==
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }

test/nettests/runnable.cpp:219: 
PASSED:
  REQUIRE( vec == expect )
with expansion:
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }
  ==
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }

[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
test/nettests/runnable.cpp:218: 
PASSED:
  REQUIRE( vec == run(false) )
with expansion:
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }
  ==
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }

test/nettests/runnable.cpp:219: 
PASSED:
  REQUIRE( vec == expect )
with expansion:
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }
  ==
  { "torproject.org", "ooni.nu", "neubot.org", "archive.org", "creativecommons
  .org", "cyber.law.harvard.edu", "duckduckgo.com", "netflix.com", "nmap.org",
  "www.emule.com" }

===============================================================================
All tests passed (9 assertions in 1 test case)

```

**Comment by bassosimone** [comment]
Tests output:

```
[!] geoip_country_path is not set
[!] geoip_asn_path is not set

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lt-runnable is a Catch v1.5.9 host application.
Run with -? for options

-------------------------------------------------------------------------------
Make sure that 'randomize_input' works
  In the common case
-------------------------------------------------------------------------------
test/nettests/runnable.cpp:167
...............................................................................

test/nettests/runnable.cpp:217: 
PASSED:
  REQUIRE( repeat(true, 8) < 8 )
with expansion:
  1 < 8

[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
[!] geoip_country_path is not set
[!] geoip_asn_path is not set
-------------------------------------------------------------------------------
Make sure that 'randomize_input' works
  When the user does not want input to be shuffled
-------------------------------------------------------------------------------
test/nettests/runnable.cpp:167
...............................................................................

test/nettests/runnable.cpp:221: 
PASSED:
  REQUIRE( repeat(false, 8) == 9 )
with expansion:
  9 == 9

===============================================================================
All tests passed (2 assertions in 1 test case)


```

**Comment by hellais** [comment]
I tested this locally and it works as expected. This is good to merge.

**Pull request merged**